### PR TITLE
Updates Formula to Support Salt 2017.7.x and 2018.3.x

### PR DESCRIPTION
**Pull request by eemperor**
These modifications fix some unicode parsing errors and address warnings related to depricated Salt modules.  There are still additional errors to resolve when using Salt 2017.7.7 and 2018.3.2.

However, Watchmaker completes successfully with these modifications when using Salt 2016.6.6 on Windows.

### CODE REVIEW

File: _modules/win_lgpo.py [hunk 1]
```diff
@@ -454,7 +454,7 @@ def policy_template_secedit(self, policies):
 
 def __virtual__():
     """Load only on Windows and only if Apply_LGPO_Delta is present."""
-    if not salt.utils.is_windows():
+    if not salt.utils.platform.is_windows():
```
Comment (lorengordon) [inline _modules/win_lgpo.py]: Looks like this change isn't working on salt 2016.11:

https://ci.appveyor.com/project/plus3it/ash-windows-formula/build/job/lv1bgf9fojxij80l?fullLog=true#L4710

```
C:\salt\salt-call.bat : [ERROR   ] Exception raised when processing __virtual__ function for win_lgpo. Module will not be loaded 'module' object has no attribute 'platform'
At line:1 char:1
+ C:\salt\salt-call.bat --local --retcode-passthrough `
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: ([ERROR   ] Exce...bute 'platform':String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
[WARNING ] salt.loaded.ext.module.win_lgpo.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'win_lgpo', please fix this.
[ERROR   ] State 'lgpo.present' was not found in SLS 'ash-windows.scm'
Reason: 'lgpo.present' is not available.
```

**Comment by eemperor** [comment]
This should be ready for a review.

Tested on Windows 2008R2, 2012R2, 2016 with Salt 2017.7.7, 2018.3.2 (both Py2 and Py3 versions).

This fix requires this update [#14](https://github.com/plus3it/watchmaker-content/pull/14) in the watchmaker-content repo.

File: _modules/win_lgpo.py [hunk 2]
```diff
@@ -541,8 +545,8 @@ def _write_policy_file(policy_template):
     tmp_ = None
     policy_file = ''
     try:
-        tmp_, policy_file = salt.utils.mkstemp(close_fd=False)
-        os.write(tmp_, os.linesep.join(policy_template))
+        tmp_, policy_file = mkstemp(close_fd=False)
+        os.write(tmp_, os.linesep.join(policy_template).encode('utf-8'))
```
Comment (lorengordon) [inline _modules/win_lgpo.py]: This is a py2/py3 type of change. We may need to be careful here. What I'll usually do these days is try to get py2 to behave more like py3... To do that, add this line before the first import statement at the top of the file, just under the docstring.

```
from __future__ import (absolute_import, division, print_function,
                        unicode_literals, with_statement)
```

Generally doing that exposes a bunch of bytes/str issues. But perhaps we'll get lucky.

And also for better py2/py3 compatibility, instead of using `os.write`, use `io.write`. Today, I'd probably rewrite this function something like this:

```
def _write_policy_file(policy_template):
    policy_file = mkstemp()
    try:
        with io.open(policy_file, mode='w') as fh_:
            fh_.write(os.linesep.join(policy_template))
    except Exception as exc:
        raise CommandExecutionError('Error saving LGPO policy file "{0}". '
                                    'Exception: {1}'.format(policy_file, exc))
    return policy_file
```

Comment (eemperor) [reply _modules/win_lgpo.py]: Okay.  Adding just the `from __future__` statement did not affect the current functionality with my latest changes.

However, incorporating `io.write` into `_write_policy_file` generates errors in processing of the policy files.

The following error is populated in some of the `err` files under `C:\Watchmaker\Logs\Ash`:
```
Format error:  expected "Computer" or "User"; found "Computer"
Policy processing aborted due to file format error
```
Watchmaker completes successfully without any indication of errors unless you inspect the `C:\Watchmaker\Logs\Ash` folder.

Taking a closer look, the error above occurs only with `REGISTRY-BASED POLICY` files.  The `SECURITY TEMPLATES` files are applied successfully.  I'll have to do some more digging to see what is different between the two policy definitions.

Comment (lorengordon) [reply _modules/win_lgpo.py]: Sorry, yeah, it's not `io.write`. That's not really a thing. The problem with coding extemporaneously. You use `io.open` to create the file handle, then use `<handle>.write(...)`. See the function rewrite from earlier as an example.

Comment (eemperor) [reply _modules/win_lgpo.py]: My bad.  I meant to say `io.open`.

Comment (eemperor) [reply _modules/win_lgpo.py]: Here's the code snippet I have:

```
def _write_policy_file(policy_template):
    policy_file = mkstemp()
    try:
        with io.open(policy_file, mode='w') as fh_:
            fh_.write(os.linesep.join(policy_template))
    except Exception as exc:
        raise CommandExecutionError('Error saving LGPO policy file "{0}". '
                                    'Exception: {1}'.format(policy_file, exc))
    return policy_file
```

Comment (lorengordon) [reply _modules/win_lgpo.py]: Alright, you were onto something with the `.encode()` bit. Something about the lgpo utility hates the file unless it is written as bytes. Here is a diff from what is currently in your branch that is at least working on Windows 2016 and with Salt PY3 2018.3.2. I think it should work on PY2 also, the way it is written, but please test.

I also added a test that will catch when the errorfile size is greater than zero, which indicates an error, even though lgpo will return 0. No more false success that way.

```diff
diff --git a/_modules/win_lgpo.py b/_modules/win_lgpo.py
index c6ca053..d9df2e3 100644
--- a/_modules/win_lgpo.py
+++ b/_modules/win_lgpo.py
@@ -16,16 +16,26 @@ their own organization.
 :depends:    Apply_LGPO_Delta.exe in %SystemRoot%\System32\
 :platform:   Windows
 """
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals, with_statement)
+
 import collections
 import logging
+import io
 import os
 import re
+
+from salt.exceptions import CommandExecutionError, SaltInvocationError
+
 try:
-    from salt.utils import mkstemp, is_windows
-except ImportError:
     from salt.utils.files import mkstemp
+except ImportError:
+    from salt.utils import mkstemp
+
+try:
     from salt.utils.platform import is_windows
-from salt.exceptions import CommandExecutionError, SaltInvocationError
+except ImportError:
+    from salt.utils import is_windows

 log = logging.getLogger(__name__)
 __virtualname__ = 'lgpo'
@@ -542,17 +552,14 @@ def validate_policies(policies):


 def _write_policy_file(policy_template):
-    tmp_ = None
-    policy_file = ''
+    policy_file = mkstemp()
     try:
-        tmp_, policy_file = mkstemp(close_fd=False)
-        os.write(tmp_, os.linesep.join(policy_template).encode('utf-8'))
-        return policy_file
+        with io.open(policy_file, mode='wb') as fh_:
+            fh_.write(os.linesep.join(policy_template).encode('utf-8'))
     except Exception as exc:
         raise CommandExecutionError('Error saving LGPO policy file "{0}". '
                                     'Exception: {1}'.format(policy_file, exc))
-    finally:
-        os.close(tmp_)
+    return policy_file


 def _write_policy_files(valid_policies):
@@ -649,6 +656,12 @@ def apply_policies(policies=None, logfile=True, errorfile=True):
         raise CommandExecutionError('Error applying LGPO policy template '
                                     '"{0}". Exception: {1}'
                                     .format(valid_policies, exc))
+
+    if errorfile and os.path.getsize(errorfile) > 0:
+        raise CommandExecutionError(
+            'Encountered errors processing the LGPO policy template. See the '
+            'error file for details -- {0}'.format(errorfile))
+
     if ret:
         raise CommandExecutionError('Non-zero exit [{0}] from {1}. We do not '
                                     'know what this means. Hopefully the '
```

Comment (lorengordon) [reply _modules/win_lgpo.py]: I went ahead and pushed a fixup commit to your branch with this change, to make it easier to review/test/incorporate...

**Comment by eemperor** [comment]
@lorengordon 

This is good to review.  Re-tested successfully with corresponding change in watchmaker-content [PR #14](https://github.com/plus3it/watchmaker-content/pull/14).

**Comment by lorengordon** [comment]
@eemperor Awesome, can you `git rebase -i master` to squash the fixup commit?
